driver-core/src/main/java/com/datastax/driver/core/Cluster.java

```diff
@@ -2002,20 +2002,38 @@ public void refreshSchemaAndSignal(final Connection connection, final DefaultRes
             if (logger.isDebugEnabled())
                 logger.debug("Refreshing schema for {}{}", keyspace == null ? "" : keyspace, table == null ? "" : '.' + table);
 
+            maybeRefreshSchemaAndSignal(connection, future, rs, keyspace, table);
+        }
+
+        public void waitForSchemaAgreementAndSignal(final Connection connection, final DefaultResultSetFuture future, final ResultSet rs) {
+            maybeRefreshSchemaAndSignal(connection, future, rs, null, null);
+        }
+
+        private void maybeRefreshSchemaAndSignal(final Connection connection, final DefaultResultSetFuture future, final ResultSet rs, final String keyspace, final String table) {
+            final boolean refreshSchema = (keyspace != null); // if false, only wait for schema agreement
+
             executor.submit(new Runnable() {
```

Comment:
Ah, yes that makes sense.  Having to handle the case where a previous waitForSchemaAgreement times out a future early into the process of something else waiting and having to resubmit is complicated.

I like the your idea.  Maybe something that can be done as part of the API cleanup for [JAVA-616](https://datastax-oss.atlassian.net/browse/JAVA-616)?